### PR TITLE
Dyna and new class structure for Classical RL

### PR DESCRIPTION
**Pull request by ajaysub110**

### CODE REVIEW

**Comment by ajaysub110** [comment]
As of now, I have implemented only the sample based Tabular Model, which can be used only for Discrete and Deterministic. There are many other models that will be useful to implement eg. GaussianModel, ExpectationModel etc.
Additionally, we could add more classical algorithms -> Monte Carlo Control, Tree Search algos (MC, TD) etc.

**Comment by ajaysub110** [comment]
@Sharad24 I was wondering if we should either move bandits out of classicalrl and into a new directory, or find a way to combine bandits with the Trainer class. What do you suggest?

**Comment by ajaysub110** [comment]
By mistake, I added the modified coverage.xml. Should we just get rid of that?

**Comment by Sharad24** [comment]
Yup. It should be generated automatically not by us.

File: jigglypuffRL/classicalrl/common/models.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        # random action in that state
+        a = np.random.choice(np.where(self.s_model[s] > 0)[0])
+        return s, a
+
+    def step(self, s, a):
+        r = self.r_model[s, a]
+        s_ = self.s_model[s, a]
+        return r, s_
+
+    def is_empty(self):
```
Comment (Sharad24) [inline jigglypuffRL/classicalrl/common/models.py:34]: Missing docstrings for methods.

File: jigglypuffRL/classicalrl/common/models.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+model_registry = {"tabular": TabularModel}
+
+
+def get_model_from_name(name_):
+    """
+    get model object from name
+    :param name_: (str) name of the model ['tabular']
+    """
+    return model_registry[name_]
```
Comment (Sharad24) [inline jigglypuffRL/classicalrl/common/models.py:46]: Check if name_ key is present in `model_registry`, I guess like we're doing for others.

File: jigglypuffRL/classicalrl/common/trainer.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            ep_r += r
+
+            if self.learning == True:
+                self.learn((s, a, r, s_))
+
+            if self.planning == True and t > self.start_plan and not self.model.is_empty():
+                self.model.add(s, a, r, s_)
+                self.plan()
+
+            s = s_
```
Comment (Sharad24) [inline jigglypuffRL/classicalrl/common/trainer.py:95]: Replace 's','a','t' with full names?

File: tests/test_classicalrl/test_common/test_trainer.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import pytest
+import numpy as np
+import gym
+
+from jigglypuffRL import SARSA, QLearning
+from jigglypuffRL.classicalrl import Trainer
+
+
+class TestTrainer:
```
Comment (Sharad24) [inline tests/test_classicalrl/test_common/test_trainer.py:9]: class not needed?

Comment (ajaysub110) [reply tests/test_classicalrl/test_common/test_trainer.py:9]: We'll add other tests in the future, so I thought it would be better to make it a class now.

Comment (Sharad24) [reply tests/test_classicalrl/test_common/test_trainer.py:9]: Ok

**Pull request merged**